libraries/abstractions/common_io/test/test_iot_sdio.c

```diff
@@ -0,0 +1,1042 @@
+    TEST_ASSERT_NOT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_card_connect(xSdioHandle);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Try to access card after card connect should succeed */
+    lRetVal = iot_sdio_io_read_direct(xSdioHandle,
+                                eSdioFunction0,
+                                testIotSdio_CMD52TESTADDR,
+                                (uint8_t *)&ucReadVal);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_card_disconnect(xSdioHandle);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Try to access card after card disconnect should fail */
+    lRetVal = iot_sdio_io_read_direct(xSdioHandle,
+                                eSdioFunction0,
+                                testIotSdio_CMD52TESTADDR,
+                                (uint8_t *)&ucReadVal);
+    TEST_ASSERT_NOT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Closing a valid handle should return success. */
+    lRetVal = iot_sdio_close(xSdioHandle);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    /* Calling card_connect() after iot_sdio_close() should fail */
+    lRetVal = iot_sdio_card_connect(xSdioHandle);
+    TEST_ASSERT_NOT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+}
+
+/**
+ * @brief Test Function to test sdio card connect and disconnect with invalid input.
+ *
+ */
+TEST( TEST_IOT_SDIO, AFQP_IotSdioCardConnectDisconnectFuzzing )
+{
+    int32_t lRetVal;
+
+    /* Should return IOT_SDIO_INVALID_VALUE on NULL handle */
+    lRetVal = iot_sdio_card_connect(NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_INVALID_VALUE, lRetVal);
+    /* Should return IOT_SDIO_INVALID_VALUE on NULL handle */
+    lRetVal = iot_sdio_card_disconnect(NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_INVALID_VALUE, lRetVal);
+}
+
+
+/**
+ * @brief Test Function to test iot_sdio_io_read_direct()
+ * iot_sdio_io_write_direct().
+ *
+ */
+TEST( TEST_IOT_SDIO, AFQP_IotSdioWriteReadDirect )
+{
+    IotSdioSlotHandle_t xSdioHandle = NULL;
+    int32_t lRetVal;
+    uint8_t ucBackupVal =0, ucReadVal =0, ucWriteVal = 0;
+    bool bCardPresent = false;
+    IoTSdioPowerOnOffCard_t xCardPowerCtl = {false, false};
+
+    /* Open with valid host index and slot index should return a non NULL handle. */
+    xSdioHandle = iot_sdio_open(uctestIotSdioValidHostIdx,
+                                uctestIotSdioValidSlotIdx);
+    TEST_ASSERT_NOT_EQUAL( NULL, xSdioHandle );
+
+    lRetVal = iot_sdio_set_sdio_callback(xSdioHandle, xtestIotSdioTestCallback, NULL);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
+
+    lRetVal = iot_sdio_ioctl(xSdioHandle, eSDIOSetCardDetectParams, (void *)&xtestIotSdioCardDetectParam);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);;
+
+    xCardPowerCtl.bPowerOn = false;
+    lRetVal = iot_sdio_ioctl(xSdioHandle,
+                             eSDIOPowerOnOffCard,
+                             (void *)&xCardPowerCtl);
+    TEST_ASSERT_EQUAL(IOT_SDIO_SUCCESS, lRetVal);
```

Comment:
What does this IOCTL validate? Is there a way to check if the IOCTL did something other than just return SUCCESS?

Reply:
Some pluggable card require card to be off state for proper detection. For embedded SDIO device, it simply return SUCCESS.

Reply:
Will update comment

Reply:
done